File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function solution(n, wires) {
    const adj = Array.from({length: n + 1}, () => []);

    for (const [from, to] of wires) {
        adj[from].push(to);
        adj[to].push(from);
    }

    function dfs(node, visited) {
        visited.add(node);
        let count = 1; 

        for (const next of adj[node]) {
            if (!visited.has(next)) {
                count += dfs(next, visited);
            }
        }
        return count;
    }

    let minDiff = n;

    for (const [from, to] of wires) {
        // 간선 제거
        adj[from] = adj[from].filter(v => v !== to);
        adj[to] = adj[to].filter(v => v !== from);

        const visited = new Set();
        const size = dfs(from, visited);
        const diff = Math.abs((n - size) - size);
        minDiff = Math.min(minDiff, diff);

        // 간선 복원
        adj[from].push(to);
        adj[to].push(from);
    }

    return minDiff;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def solution(n, wires):
    graph = [[0] * (n+1) for _ in range(n+1)]
    
    for a, b in wires:
        graph[a][b] = graph[b][a] = 1
    
    minDiff = n
    
    def dfs(r, c):
        cnt[0] += 1
        visited[r][c] = visited[c][r] = True
        
        for i, nxt in enumerate(graph[c]):
            if nxt == 1 and not visited[c][i]:
                dfs(c, i)
    
    for r in range(n+1):
        for c in range(n+1):
            if graph[r][c] == 0:
                continue
            graph[r][c] = graph[c][r] = 0
            visited = [[False] * (n+1) for _ in range(n+1)]
            cnt = [0]
            dfs(r, c)
            minDiff = min(minDiff, abs(2 * cnt[0] -n))
            graph[r][c] = graph[c][r] = 1
            
    return minDiff